### PR TITLE
Ticket/305

### LINKED ISSUE #305: Track fees on payments & allow more detailed viewing of payments
This is related to #215 and #296. In order to give an accurate financial overview of the event (and provide money-related analytics) we should keep track of how much is going to our fees vs. payment provider fees. This *should* be accessible via dwolla & stripe APIs.

Checklist:

* [x] Change models to support processing / application fee storage
* [x] Add support for Stripe fees on purchase
* [x] Add support for Dwolla fees on purchase
* [x] Add support for Stripe fees on refund
* [x] Add support for Dwolla fees on refund
* [x] Add tests

### PR DESCRIPTION
**Pull request by melinath**
This resolves #305 by merging Payment and Refund into a single Transaction model that additionally tracks any associated application / payment processing fees. This even has tests! There are some additional enhancements we'll want to make to the finances page for organizers - entering of arbitrary data, display of pie charts, and others - but these are separate tickets and can/should be handled later.

Even though this does have tests, I would appreciate a run-through to make sure everything seems fine.

### CODE REVIEW

**Comment by harrislapiroff** [comment]
Not sure if this is new or not. Got while attempting to pay for a pass via Stripe with a previously saved card.

```pytb
Environment:


Request Method: POST
Request URL: http://localhost:8000/new-dance-weekend/order/summary/

Django Version: 1.7.4
Python Version: 2.7.6
Installed Applications:
('brambling',
 'grappelli',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'zenaida',
 'zenaida.contrib.hints',
 'floppyforms',
 'django_filters',
 'daguerre',
 'debug_toolbar.apps.DebugToolbarConfig',
 'template_timings_panel',
 'zenaida.contrib.feedback')
Installed Middleware:
(u'debug_toolbar.middleware.DebugToolbarMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'zenaida.contrib.feedback.middleware.FeedbackMiddleware')


Traceback:
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  111.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/views/generic/base.py" in view
  69.             return self.dispatch(request, *args, **kwargs)
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/views/orders.py" in dispatch
  258.         return super(OrderMixin, self).dispatch(request, *args, **kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/views/generic/base.py" in dispatch
  87.         return handler(request, *args, **kwargs)
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/views/orders.py" in post
  776.             if form and form.is_valid():
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/forms/forms.py" in is_valid
  162.         return self.is_bound and not bool(self.errors)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/forms/forms.py" in errors
  154.             self.full_clean()
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/forms/forms.py" in full_clean
  355.         self._post_clean()
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/forms/orders.py" in _post_clean
  463.                 event=self.event,

Exception Type: AttributeError at /new-dance-weekend/order/summary/
Exception Value: 'SavedCardPaymentForm' object has no attribute 'event'
```

**Comment by harrislapiroff** [comment]
Ditto for the `OneTimePaymentForm`:

```pytb
Environment:


Request Method: POST
Request URL: http://localhost:8000/new-dance-weekend/order/summary/

Django Version: 1.7.4
Python Version: 2.7.6
Installed Applications:
('brambling',
 'grappelli',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'zenaida',
 'zenaida.contrib.hints',
 'floppyforms',
 'django_filters',
 'daguerre',
 'debug_toolbar.apps.DebugToolbarConfig',
 'template_timings_panel',
 'zenaida.contrib.feedback')
Installed Middleware:
(u'debug_toolbar.middleware.DebugToolbarMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'zenaida.contrib.feedback.middleware.FeedbackMiddleware')


Traceback:
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  111.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/views/generic/base.py" in view
  69.             return self.dispatch(request, *args, **kwargs)
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/views/orders.py" in dispatch
  258.         return super(OrderMixin, self).dispatch(request, *args, **kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/views/generic/base.py" in dispatch
  87.         return handler(request, *args, **kwargs)
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/views/orders.py" in post
  776.             if form and form.is_valid():
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/forms/forms.py" in is_valid
  162.         return self.is_bound and not bool(self.errors)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/forms/forms.py" in errors
  154.             self.full_clean()
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/forms/forms.py" in full_clean
  355.         self._post_clean()
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/forms/orders.py" in _post_clean
  423.             'event': self.event,

Exception Type: AttributeError at /new-dance-weekend/order/summary/
Exception Value: 'OneTimePaymentForm' object has no attribute 'event'
```

**Comment by harrislapiroff** [comment]
Also, not sure if this is new, but every time I save event settings, it disconnects my Stripe payment.

**Comment by harrislapiroff** [comment]
Also got this one while paying by check. Not sure if it's a logic error or an error in migrations. The event existed before I ran migrations, but I had not attempted to place an order until after the migrations:

```pytb
Environment:


Request Method: POST
Request URL: http://localhost:8000/new-dance-weekend/order/summary/

Django Version: 1.7.4
Python Version: 2.7.6
Installed Applications:
('brambling',
 'grappelli',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'zenaida',
 'zenaida.contrib.hints',
 'floppyforms',
 'django_filters',
 'daguerre',
 'debug_toolbar.apps.DebugToolbarConfig',
 'template_timings_panel',
 'zenaida.contrib.feedback')
Installed Middleware:
(u'debug_toolbar.middleware.DebugToolbarMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'zenaida.contrib.feedback.middleware.FeedbackMiddleware')


Traceback:
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  111.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/views/generic/base.py" in view
  69.             return self.dispatch(request, *args, **kwargs)
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/views/orders.py" in dispatch
  258.         return super(OrderMixin, self).dispatch(request, *args, **kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/views/generic/base.py" in dispatch
  87.         return handler(request, *args, **kwargs)
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/views/orders.py" in post
  777.                 payment = form.save()
File "/Users/harris/Projects/Clients/Little Weaver/django-brambling/brambling/forms/orders.py" in save
  509.             api_type=self.api_type
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/manager.py" in manager_method
  92.                 return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/query.py" in create
  372.         obj.save(force_insert=True, using=self.db)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/base.py" in save
  589.                        force_update=force_update, update_fields=update_fields)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/base.py" in save_base
  617.             updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/base.py" in _save_table
  698.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/base.py" in _do_insert
  731.                                using=using, raw=raw)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/manager.py" in manager_method
  92.                 return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/query.py" in _insert
  921.         return query.get_compiler(using=using).execute_sql(return_id)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  920.                 cursor.execute(sql, params)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/debug_toolbar/panels/sql/tracking.py" in execute
  174.         return self._record(self.cursor.execute, sql, params)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/debug_toolbar/panels/sql/tracking.py" in _record
  104.             return method(sql, params)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  81.             return super(CursorDebugWrapper, self).execute(sql, params)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  65.                 return self.cursor.execute(sql, params)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/utils.py" in __exit__
  94.                 six.reraise(dj_exc_type, dj_exc_value, traceback)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  65.                 return self.cursor.execute(sql, params)
File "/Users/harris/.virtualenvs/brambling/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py" in execute
  485.         return Database.Cursor.execute(self, query, params)

Exception Type: IntegrityError at /new-dance-weekend/order/summary/
Exception Value: NOT NULL constraint failed: brambling_transaction.event_id
```

**Comment by harrislapiroff** [comment]
Tried it on a newly created event and got the same error.

**Comment by harrislapiroff** [comment]
TLDR:

* [ ] `SavedCardPaymentForm`: 'SavedCardPaymentForm' object has no attribute 'event' at `/brambling/forms/orders.py` line 463
* [ ] Ditto for `OneTimePaymentForm`
* [ ] Bug where Stripe Connect gets disconnected every time I save event settings
* [ ] Can't complete checkout due to `IntegrityError`

**Comment by melinath** [comment]
The third item (stripe and dwolla getting disconnected) is due to a separate bug that probably also exists on master: The test api fields are not excluded from the EventForm. So test events will be reset constantly. Relatedly, test events *won't* be able to choose to disconnect their accounts. Also relatedly, we should *actually* disconnect people, not just erase our record of their data.

**Comment by melinath** [comment]
Thanks, Harris! Fixed & tested.

**Comment by melinath** [comment]
Which is to say, I added tests for the parts that you saw breaking. I suppose there could hypothetically be other broken bits, but I think I hit all the parts that had changed since the last time I actually used the forms...

**Comment by harrislapiroff** [comment]
I didn't run into anything else broken. Merging.
